discovery-provider/src/solana/anchor_program_indexer.py

```diff
@@ -0,0 +1,170 @@
+import asyncio
+import base64
+import json
+import logging
+from pathlib import Path
+from typing import Any, Dict, List, Optional
+
+import base58
+from anchorpy import Idl, InstructionCoder
+from solana.transaction import Transaction, TransactionInstruction
+from sqlalchemy import desc
+from src.models.models import AudiusDataTx
+from src.solana.solana_program_indexer import SolanaProgramIndexer
+from src.utils.helpers import split_list
+
+logger = logging.getLogger(__name__)
+
+BASE_ERROR = "Must be implemented in subclass"
+TX_SIGNATURES_PROCESSING_SIZE = 100
+AUDIUS_DATA_IDL_PATH = "./idl/audius_data.json"
+
+
+class AnchorDataIndexer(SolanaProgramIndexer):
+    """
+    Indexer for the audius user data layer
+    """
+
+    _instruction_coder: InstructionCoder
+
+    def __init__(
+        self,
+        program_id: str,
+        label: str,
+        redis: Any,
+        db: Any,
+        solana_client_manager: Any,
+    ):
+        super().__init__(program_id, label, redis, db, solana_client_manager)
+        self._init_instruction_coder()
+
+    def get_tx_in_db(self, session: Any, tx_sig: str):
+        exists = False
+        tx_sig_db_count = (
+            session.query(AudiusDataTx).filter(AudiusDataTx.signature == tx_sig)
+        ).count()
+        exists = tx_sig_db_count > 0
+        self.msg(f"{tx_sig} exists={exists}")
+        return exists
+
+    def get_latest_slot(self):
+        latest_slot = None
+        with self._db.scoped_session() as session:
+            highest_slot_query = (
+                session.query(AudiusDataTx)
+                .filter(AudiusDataTx.slot != None)
+                .filter(AudiusDataTx.signature != None)
+                .order_by(desc(AudiusDataTx.slot))
+            ).first()
+            # Can be None prior to first write operations
+            if highest_slot_query is not None:
+                latest_slot = highest_slot_query.slot
+
+        # If no slots have yet been recorded, assume all are valid
+        if latest_slot is None:
+            latest_slot = 0
+
+        self.msg(f"returning {latest_slot} for highest slot")
+        return latest_slot
+
+    def validate_and_save_parsed_tx_records(
+        self, processed_transactions, metadata_dictionary
+    ):
+        self.msg(
+            f"validate_and_save anchor {processed_transactions} - {metadata_dictionary}"
+        )
+        with self._db.scoped_session() as session:
+            for transaction in processed_transactions:
+                session.add(
+                    AudiusDataTx(
+                        signature=transaction["tx_sig"],
+                        slot=transaction["result"]["slot"],
+                    )
+                )
+
+    # TODO - This is where we will deserialize instruction data and accounts
+    # tx_metadata should contain TxType, Deserialized Instruction Data, etc
+    def parse_tx(self, tx_sig):
```

Comment:
could be its own class if it makes testing easier.

Reply:
Moved sub functions into their own module for isolated testing